simpleast-core/src/main/java/com/discord/simpleast/code/JavaScript.kt

```diff
@@ -0,0 +1,169 @@
+    "Error|SyntaxError|TypeError|RangeError|ReferenceError|EvalError|AggregateError"
+  )
+
+  class FunctionNode<RC>(
+    pre: String?, signature: String?, params: String,
+    codeStyleProviders: CodeStyleProviders<RC>
+  ) : Node.Parent<RC>(
+      pre?.let { StyleNode.TextStyledNode(pre, codeStyleProviders.keywordStyleProvider) },
+      signature?.let { StyleNode.TextStyledNode(signature, codeStyleProviders.identifierStyleProvider) },
+      StyleNode.TextStyledNode(params, codeStyleProviders.paramsStyleProvider)
+  ) : {
+      companion object {
+        /**
+         * Matches against a JavaScript function declaration.
+         *
+         * ```
+         * function foo(bar)
+         * function baz()
+         * ```
+         */
+         private val PATTERN_JAVASCRIPT_FUNC = 
+             """^(function\*?|static|get|set)? *?(\w+)?( *?\(.*?\)) *?\{[\s\S]*?\}""".toRegex(RegexOption.DOT_MATCH_ALL).toPattern()
+
+         fun <RC, S> createFunctionRule(codeStyleProviders: CodeStyleProviders<RC>) =
+          object : Rule<RC, Node<RC>, S>(PATTERN_JAVASCRIPT_FUNC) {
+            override fun parse(matcher: Matcher, parser: Parser<RC, in Node<RC>, S>, state: S): ParseSpec<RC, S> {
+              val definition = matcher.group(1)
+              val signature = matcher.group(2)
+              val params = matcher.group(3)
+              return ParseSpec.createTerminal(FunctionNode(definition, signature, params!!, codeStyleProviders), state)
+            }
+          }
+    }
+  }
+
+  class FieldNode<RC>(
+      definition: String, name: String,
+      codeStyleProviders: CodeStyleProviders<RC>
+  ) : Node.Parent<RC>(
+      StyleNode.TextStyledNode(definition, codeStyleProviders.keywordStyleProvider),
+      StyleNode.TextStyledNode(name, codeStyleProviders.identifierStyleProvider),
+  ) {
+    companion object {
+      /**
+       * Matches against a JavaScript field definition.
+       *
+       * ```
+       * var x = 1;
+       * let y = 5;
+       * const z = 10;
+       * ```
+       */
+      private val PATTERN_JAVASCRIPT_FIELD =
+          Pattern.compile("""^(var|let|const)\s+(?:\{\s*?(.*)\s*?\}|(\w+))""", Pattern.DOTALL)
+
+      fun <RC, S> createFieldRule(
+          codeStyleProviders: CodeStyleProviders<RC>
+      ) =
+          object : Rule<RC, Node<RC>, S>(PATTERN_JAVASCRIPT_FIELD) {
+            override fun parse(matcher: Matcher, parser: Parser<RC, in Node<RC>, S>, state: S):
+                ParseSpec<RC, S> {
+              val definition = matcher.group(1)
+              val name = matcher.group(3)
+              if (!name) name = matcher.group(2)
+              return ParseSpec.createTerminal(
+                  FieldNode(definition!!, name!!, codeStyleProviders), state)
+            }
+          }
+    }
+  }
+
+  /**
+   * Matches against a JavaScript regex.
+   *
+   * ```
+   * /(.*)/
+   * ```
+   */
+   private val PATTERN_JAVASCRIPT_REGEX = 
+       Pattern.compile("""^/.*?/\w+?""", Pattern.DOTALL)
```

Comment:
```suggestion
       Pattern.compile("""^/.*?/(?:\w*)?""", Pattern.DOTALL)
```
I'm assuming you meant to make the RegExp flags optional.

Reply:
Yep, missed that, thanks for pointing that out, fixed.